File: app/src/main/java/com/example/minesweeper/MainActivity.java
package com.example.minesweeper;

import android.os.Bundle;
import android.os.SystemClock;
import android.view.View;
import android.widget.Button;
import android.widget.Chronometer;
import android.widget.GridLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

public class MainActivity extends AppCompatActivity implements GameEngine.GameListener {
    private GridLayout gridLayout;
    private GameEngine gameEngine;
    private Button[][] cellButtons;
    private TextView minesCountText;
    private TextView statusText;
    private Chronometer timer;
    private Button restartButton;
    private Button difficultyButton;
    private int currentDifficulty = 0; // 0: Easy, 1: Medium, 2: Hard
    private boolean timerRunning = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        gridLayout = findViewById(R.id.gridLayout);
        minesCountText = findViewById(R.id.minesCountText);
        statusText = findViewById(R.id.statusText);
        timer = findViewById(R.id.timer);
        restartButton = findViewById(R.id.restartButton);
        difficultyButton = findViewById(R.id.difficultyButton);

        restartButton.setOnClickListener(v -> startNewGame());
        difficultyButton.setOnClickListener(v -> showDifficultyDialog());

        startNewGame();
    }

    private void startNewGame() {
        gameEngine = new GameEngine(currentDifficulty);
        gameEngine.setListener(this);
        createGrid();
        updateMinesCount();
        statusText.setText(getString(R.string.game_status_playing));
        timerRunning = false;
        timer.stop();
        timer.setBase(SystemClock.elapsedRealtime());
    }

    private void createGrid() {
        gridLayout.removeAllViews();
        int rows = gameEngine.getRows();
        int cols = gameEngine.getCols();
        
        gridLayout.setRowCount(rows);
        gridLayout.setColumnCount(cols);
        
        cellButtons = new Button[rows][cols];
        
        int cellSize = getResources().getDisplayMetrics().widthPixels / cols - 10;
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                Button button = new Button(this);
                button.setLayoutParams(new GridLayout.LayoutParams(
                    GridLayout.spec(i, 1f),
                    GridLayout.spec(j, 1f)
                ));
                button.setMinWidth(cellSize);
                button.setMinHeight(cellSize);
                button.setTextSize(12);
                button.setPadding(0, 0, 0, 0);
                
                final int row = i;
                final int col = j;
                
                button.setOnClickListener(v -> {
                    if (!timerRunning) {
                        timer.setBase(SystemClock.elapsedRealtime());
                        timer.start();
                        timerRunning = true;
                    }
                    gameEngine.revealCell(row, col);
                });
                
                button.setOnLongClickListener(v -> {
                    gameEngine.toggleFlag(row, col);
                    updateCell(row, col);
                    return true;
                });
                
                cellButtons[i][j] = button;
                gridLayout.addView(button);
                updateCell(i, j);
            }
        }
    }

    private void updateCell(int row, int col) {
        Button button = cellButtons[row][col];
        Cell cell = gameEngine.getCell(row, col);
        
        if (cell.isFlagged()) {
            button.setText("🚩");
            button.setBackgroundColor(ContextCompat.getColor(this, android.R.color.holo_orange_light));
        } else if (cell.isRevealed()) {
            if (cell.isMine()) {
                button.setText("💣");
                button.setBackgroundColor(ContextCompat.getColor(this, android.R.color.holo_red_light));
            } else {
                int adjacentMines = cell.getAdjacentMines();
                if (adjacentMines > 0) {
                    button.setText(String.valueOf(adjacentMines));
                    button.setTextColor(getNumberColor(adjacentMines));
                } else {
                    button.setText("");
                }
                button.setBackgroundColor(ContextCompat.getColor(this, android.R.color.darker_gray));
            }
        } else {
            button.setText("");
            button.setBackgroundColor(ContextCompat.getColor(this, android.R.color.background_light));
        }
    }

    private int getNumberColor(int number) {
        switch (number) {
            case 1: return ContextCompat.getColor(this, android.R.color.holo_blue_dark);
            case 2: return ContextCompat.getColor(this, android.R.color.holo_green_dark);
            case 3: return ContextCompat.getColor(this, android.R.color.holo_red_dark);
            case 4: return ContextCompat.getColor(this, android.R.color.holo_purple);
            case 5: return ContextCompat.getColor(this, android.R.color.holo_orange_dark);
            case 6: return ContextCompat.getColor(this, android.R.color.holo_red_light);
            case 7: return ContextCompat.getColor(this, android.R.color.black);
            case 8: return ContextCompat.getColor(this, android.R.color.darker_gray);
            default: return ContextCompat.getColor(this, android.R.color.black);
        }
    }

    private void updateMinesCount() {
        int remaining = gameEngine.getTotalMines() - getFlaggedCount();
        minesCountText.setText(String.format("雷: %d", remaining));
    }

    private int getFlaggedCount() {
        int count = 0;
        for (int i = 0; i < gameEngine.getRows(); i++) {
            for (int j = 0; j < gameEngine.getCols(); j++) {
                if (gameEngine.getCell(i, j).isFlagged()) {
                    count++;
                }
            }
        }
        return count;
    }

    private void showDifficultyDialog() {
        String[] difficulties = {"简单 (9x9, 10雷)", "中等 (16x16, 40雷)", "困难 (16x30, 99雷)"};
        new AlertDialog.Builder(this)
            .setTitle("选择难度")
            .setItems(difficulties, (dialog, which) -> {
                currentDifficulty = which;
                startNewGame();
            })
            .show();
    }

    @Override
    public void onCellRevealed(int row, int col) {
        runOnUiThread(() -> {
            updateCell(row, col);
            if (gameEngine.isGameOver() || gameEngine.isGameWon()) {
                updateAllCells();
            }
        });
    }

    @Override
    public void onGameOver(boolean won) {
        runOnUiThread(() -> {
            timer.stop();
            timerRunning = false;
            if (won) {
                statusText.setText(getString(R.string.game_status_won));
                Toast.makeText(this, "恭喜！你赢了！", Toast.LENGTH_LONG).show();
            } else {
                statusText.setText(getString(R.string.game_status_lost));
                gameEngine.revealAllMines();
                updateAllCells();
                Toast.makeText(this, "游戏结束！", Toast.LENGTH_LONG).show();
            }
        });
    }

    @Override
    public void onMinesCountChanged(int remainingMines) {
        runOnUiThread(() -> updateMinesCount());
    }

    private void updateAllCells() {
        for (int i = 0; i < gameEngine.getRows(); i++) {
            for (int j = 0; j < gameEngine.getCols(); j++) {
                updateCell(i, j);
            }
        }
    }
}

File: app/src/main/java/com/example/minesweeper/MainActivity.kt
package com.example.minesweeper

import android.os.Bundle
import android.os.SystemClock
import android.widget.Button
import android.widget.Chronometer
import android.widget.GridLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat

class MainActivity : AppCompatActivity(), GameEngine.GameListener {
    private lateinit var gridLayout: GridLayout
    private lateinit var gameEngine: GameEngine
    private lateinit var cellButtons: Array<Array<Button>>
    private lateinit var minesCountText: TextView
    private lateinit var statusText: TextView
    private lateinit var timer: Chronometer
    private lateinit var restartButton: Button
    private lateinit var difficultyButton: Button
    private var currentDifficulty = 0 // 0: Easy, 1: Medium, 2: Hard
    private var timerRunning = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        gridLayout = findViewById(R.id.gridLayout)
        minesCountText = findViewById(R.id.minesCountText)
        statusText = findViewById(R.id.statusText)
        timer = findViewById(R.id.timer)
        restartButton = findViewById(R.id.restartButton)
        difficultyButton = findViewById(R.id.difficultyButton)

        restartButton.setOnClickListener { startNewGame() }
        difficultyButton.setOnClickListener { showDifficultyDialog() }

        startNewGame()
    }

    private fun startNewGame() {
        gameEngine = GameEngine(currentDifficulty)
        gameEngine.setListener(this)
        createGrid()
        updateMinesCount()
        statusText.setText(R.string.game_status_playing)
        timerRunning = false
        timer.stop()
        timer.base = SystemClock.elapsedRealtime()
    }

    private fun createGrid() {
        gridLayout.removeAllViews()
        val rows = gameEngine.getRows()
        val cols = gameEngine.getCols()

        gridLayout.rowCount = rows
        gridLayout.columnCount = cols

        cellButtons = Array(rows) { Array(cols) { Button(this) } }

        val cellSize = resources.displayMetrics.widthPixels / cols - 10

        for (i in 0 until rows) {
            for (j in 0 until cols) {
                val button = cellButtons[i][j]
                button.layoutParams = GridLayout.LayoutParams(
                    GridLayout.spec(i, 1f),
                    GridLayout.spec(j, 1f)
                )
                button.minWidth = cellSize
                button.minHeight = cellSize
                button.textSize = 12f
                button.setPadding(0, 0, 0, 0)

                val row = i
                val col = j

                button.setOnClickListener {
                    if (!timerRunning) {
                        timer.base = SystemClock.elapsedRealtime()
                        timer.start()
                        timerRunning = true
                    }
                    gameEngine.revealCell(row, col)
                }

                button.setOnLongClickListener {
                    gameEngine.toggleFlag(row, col)
                    updateCell(row, col)
                    true
                }

                gridLayout.addView(button)
                updateCell(i, j)
            }
        }
    }

    private fun updateCell(row: Int, col: Int) {
        val button = cellButtons[row][col]
        val cell = gameEngine.getCell(row, col)

        when {
            cell.isFlagged -> {
                button.text = "🚩"
                button.setBackgroundColor(ContextCompat.getColor(this, android.R.color.holo_orange_light))
            }
            cell.isRevealed -> {
                if (cell.isMine) {
                    button.text = "💣"
                    button.setBackgroundColor(ContextCompat.getColor(this, android.R.color.holo_red_light))
                } else {
                    if (cell.adjacentMines > 0) {
                        button.text = cell.adjacentMines.toString()
                        button.setTextColor(getNumberColor(cell.adjacentMines))
                    } else {
                        button.text = ""
                    }
                    button.setBackgroundColor(ContextCompat.getColor(this, android.R.color.darker_gray))
                }
            }
            else -> {
                button.text = ""
                button.setBackgroundColor(ContextCompat.getColor(this, android.R.color.background_light))
            }
        }
    }

    private fun getNumberColor(number: Int): Int {
        return when (number) {
            1 -> ContextCompat.getColor(this, android.R.color.holo_blue_dark)
            2 -> ContextCompat.getColor(this, android.R.color.holo_green_dark)
            3 -> ContextCompat.getColor(this, android.R.color.holo_red_dark)
            4 -> ContextCompat.getColor(this, android.R.color.holo_purple)
            5 -> ContextCompat.getColor(this, android.R.color.holo_orange_dark)
            6 -> ContextCompat.getColor(this, android.R.color.holo_red_light)
            7 -> ContextCompat.getColor(this, android.R.color.black)
            8 -> ContextCompat.getColor(this, android.R.color.darker_gray)
            else -> ContextCompat.getColor(this, android.R.color.black)
        }
    }

    private fun updateMinesCount() {
        val remaining = gameEngine.getTotalMines() - getFlaggedCount()
        minesCountText.text = "雷: $remaining"
    }

    private fun getFlaggedCount(): Int {
        var count = 0
        for (i in 0 until gameEngine.getRows()) {
            for (j in 0 until gameEngine.getCols()) {
                if (gameEngine.getCell(i, j).isFlagged) {
                    count++
                }
            }
        }
        return count
    }

    private fun showDifficultyDialog() {
        val difficulties = arrayOf("简单 (9x9, 10雷)", "中等 (16x16, 40雷)", "困难 (16x30, 99雷)")
        AlertDialog.Builder(this)
            .setTitle("选择难度")
            .setItems(difficulties) { _, which ->
                currentDifficulty = which
                startNewGame()
            }
            .show()
    }

    override fun onCellRevealed(row: Int, col: Int) {
        runOnUiThread {
            updateCell(row, col)
            if (gameEngine.isGameOver() || gameEngine.isGameWon()) {
                updateAllCells()
            }
        }
    }

    override fun onGameOver(won: Boolean) {
        runOnUiThread {
            timer.stop()
            timerRunning = false
            if (won) {
                statusText.setText(R.string.game_status_won)
                Toast.makeText(this, "恭喜！你赢了！", Toast.LENGTH_LONG).show()
            } else {
                statusText.setText(R.string.game_status_lost)
                gameEngine.revealAllMines()
                updateAllCells()
                Toast.makeText(this, "游戏结束！", Toast.LENGTH_LONG).show()
            }
        }
    }

    override fun onMinesCountChanged(remainingMines: Int) {
        runOnUiThread { updateMinesCount() }
    }

    private fun updateAllCells() {
        for (i in 0 until gameEngine.getRows()) {
            for (j in 0 until gameEngine.getCols()) {
                updateCell(i, j)
            }
        }
    }
}

